lib/rust/ensogl/core/src/display/object/instance.rs

```diff
@@ -1020,133 +1194,862 @@ impl InstanceDef {
+    /// updated in the horizontal direction, then in the vertical direction. To better illustrate
+    /// the need of such a solution, consider the following example:
+    ///
+    /// ```text
+    /// ╭▷ ROOT ──────────────────────────╮
+    /// │   ╭▷ L ◀ ▶ ──╮   ╭R─ ▶ ◀ ───╮   │   Auto-layout Legend:         
+    /// │   │ ╭ ◀ ▶ ╮  │   ▽ ╭────╮   │   │   ┄── ▷ ──┄ : Horizontal auto-layout.
+    /// │   │ │ L1  │  │   │ │ R1 ▲   │   │   ┄── ▽ ──┄ : Vertical auto-layout.
+    /// │   │ │     │  │   │ │    ▼   │   │   ┄───────┄ : Manual layout.   
+    /// │   │ │     │  ▼   │ ╰────╯   ▲   │
+    /// │   │ │     │  ▲   │ ╭────╮   ▼   │   Resizing Legend:             
+    /// │   │ │     │  │   │ │ R2 ▲   │   │   ┄── ◀ ▶ ──┄ : Fill resizing.
+    /// │   │ │     │  │   │ │    ▼   │   │   ┄── ▶ ◀ ──┄ : Hug resizing.  
+    /// │   │ ╰─────╯  │   │ ╰────╯   │   │   ┄─────────┄ : Fixed resizing.
+    /// │   ╰──────────╯   ╰──────────╯   │
+    /// ╰─────────────────────────────────╯
+    /// ```
+    ///
+    /// 1. In the first pass, we are updating the horizontal layout.
+    ///    a) First, we are visiting the `L` object. It's X-axis resizing is set to `Fill`, so we
+    ///       can't determine its width yet. Neither we can update the X-axis layout of its child,
+    ///       as it may depend on the `L` object width.
+    ///    b) Then, we are visiting the `R` object. It's X-axis resizing is set to `Hug`, so we need
+    ///       to visit its children to find the widest one. It's layout is set to vertical. Unlike
+    ///       in the case of the `L` and `R` objects, we are computing the size in the orthogonal
+    ///       direction than the layout the children are placed in.
+    ///    c) As the `ROOT` object's width resizing is set to `Fixed`, after finding the `R` object
+    ///       width, we can compute the `L` object width.
+    ///    d) Finally, we can update the `L` object children layout.
+    ///
+    /// 2. In the second pass, we are updating the vertical layout.
+    ///    a) First, we are visiting the `L` object. It's Y-axis resizing is set to `Hug`, so we
+    ///       need to visit its children to find the tallest one.
+    ///    b) The `L1` object's Y-axis resizing is set to `Fixed`, so we can simply update its
+    ///       height.
+    ///    c) Next, we are visiting the `R` object. It's Y-axis resizing is set to `Fill`, so we can
+    ///       compute it, as the `ROOT` object height is fixed.
+    ///    d) Finally, we can update the `R` object children layout. Both children Y-axis resizing
+    ///       is set to `Fill`, so they are equally using the available space.
+    ///
+    /// Please note, that this algorithm could not be realized in a single pass, as we can't compute
+    /// `L` object width without first computing the `R` object width, and we can't compute the
+    /// `R` object height without first computing the `L` object height.
+    ///
```

Comment:
But that's not true for this example: the R height is Fill and ROOT height is fixed, so we can just set height to ROOT's height, can't we? The L's height does not seem to affect anything.

Reply:
I described it better.